### PR TITLE
*(ticdc): Initialize changefeed/processor asynchronously

### LINKED ISSUE #10845: Make TiCDC initialize changefeed asynchronously.
As indicated by #10789 and #10797 the ticdc owner and processor may experience significant delays during a cdc server restart if there are numerous changefeeds. This happens because each changefeed is initialized synchronously, one after the other.

To mitigate this issue, we should incorporate logic to initialize the changefeed asynchronously.

### PR DESCRIPTION
**Pull request by sdojjy**
### What problem does this PR solve?

Issue Number: close #10845

### What is changed and how it works?
1. create and run a aysnc thread pool when reset capture, the thread pool size is 8, and it's hardcoded.
2. init changefeed and processor using the async thread pool

CDC rolling upgrade LAG is less than 10s

image

#### Tests 

 - Unit test
 - Integration test
 - Manual test (add detailed scripts or steps below)

### Release note 

```release-note
`None`.
```

### CODE REVIEW

File: cdc/owner/changefeed.go [hunk 1]
```diff
@@ -376,9 +379,22 @@ func (c *changefeed) tick(ctx cdcContext.Context,
 	if adminJobPending {
 		return 0, 0, nil
 	}
+	if !c.initializing && !c.initialized {
+		c.initializing = true
```
Comment (asddongmen) [inline cdc/owner/changefeed.go:383]: initializing and initialized should be atomic values

**Review by asddongmen — APPROVED** [review]

File: cdc/async/async_initializer.go [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+	"github.com/pingcap/tiflow/pkg/workerpool"
+	"go.uber.org/atomic"
+	"go.uber.org/zap"
+)
+
+// Initializer is a helper struct to initialize a changefeed asynchronously.
+type Initializer struct {
+	// state related fields
+	initialized  *atomic.Bool
+	initializing *atomic.Bool
```
Comment (3AceShowHand) [inline cdc/async/async_initializer.go:31]: It looks define some state to indicate the initialize phase is better.

File: cdc/async/async_initializer.go [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+
+	"github.com/pingcap/log"
+	"github.com/pingcap/tiflow/pkg/errors"
+	"github.com/pingcap/tiflow/pkg/workerpool"
+	"go.uber.org/atomic"
+	"go.uber.org/zap"
+)
+
+// Initializer is a helper struct to initialize a changefeed asynchronously.
+type Initializer struct {
```
Comment (3AceShowHand) [inline cdc/async/async_initializer.go:28]: IMHO, the name of variables can be simplified further. the prefix `initializ` is spread all over the file, hard to read.

File: cdc/async/async_initializer.go [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+func (initializer *Initializer) TryInitialize(ctx context.Context, pool workerpool.AsyncPool) (bool, error) {
+	if initializer.initialized.Load() {
+		return true, nil
+	}
+	if initializer.initializing.CompareAndSwap(false, true) {
+		initialCtx, cancelInitialize := context.WithCancel(ctx)
+		initializer.initialWaitGroup.Add(1)
+		initializer.cancelInitialize = cancelInitialize
+		log.Info("submit async initializer task to the worker pool")
+		err := pool.Go(initialCtx, func() {
```
Comment (3AceShowHand) [inline cdc/async/async_initializer.go:66]: When this error may happens, what to do next ?

**Review by 3AceShowHand — APPROVED** [review]

**Pull request merged**